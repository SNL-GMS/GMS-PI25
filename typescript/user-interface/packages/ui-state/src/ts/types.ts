import type {
  ChannelSegmentTypes,
  ChannelTypes,
  CommonTypes,
  EventTypes,
  FilterTypes,
  FkTypes,
  SignalDetectionTypes
} from '@gms/common-model';
import type { AssociationConflict, Event } from '@gms/common-model/lib/event';
import type {
  FilterDefinition,
  FilterDefinitionByFilterDefinitionUsage
} from '@gms/common-model/lib/filter/types';
import type { QcSegment } from '@gms/common-model/lib/qc-segment';
import type { SignalDetectionHypothesis } from '@gms/common-model/lib/signal-detection';
import type { Station } from '@gms/common-model/lib/station-definitions/station-definitions/station-definitions';

export interface UiChannelSegment {
  channelSegmentDescriptor: ChannelSegmentTypes.ChannelSegmentDescriptor;
  processingMasks: ChannelSegmentTypes.ProcessingMask[];
  channelSegment: ChannelSegmentTypes.ChannelSegment<ChannelSegmentTypes.TimeSeries>;
  domainTimeRange?: CommonTypes.TimeRange;
}

export type ChannelName = UiChannelSegment['channelSegmentDescriptor']['channel']['name'];
export type StationName = Station['name'];
/** Weavess row could be a channel name or station name */
export type WeavessRowName = ChannelName | StationName;
export type FilterName = FilterDefinition['name'];
/**
 * A combination of the namedFilter name (if it exists) + filter definition name
 */
export type FilterNameId = string;
export type SampleRate = FilterDefinition['filterDescription']['parameters']['sampleRateHz'];
/**
 * JSON.stringify({@link ChannelSegmentTypes.ChannelSegmentDescriptor})
 */
export type ChannelSegmentDescriptorId = string;
export type SignalDetectionIdString = SignalDetectionTypes.SignalDetection['id'];
export type SignalDetectionHypothesisId = SignalDetectionHypothesis['id']['id'];
export type EventId = Event['id'];
export type QcSegmentId = QcSegment['id'];

/**
 * Record<{@link ChannelName}, Record<{@link FilterName}, {@link UiChannelSegment}[]>>
 *
 * A record of channel names, each containing a record of filter names to an array of ui channel segments.
 */
export type UIChannelSegmentRecord = Record<ChannelName, Record<FilterName, UiChannelSegment[]>>;

/**
 * Record<JSON.stringify({@link ChannelSegmentTypes.ChannelSegmentDescriptor}), {@link ChannelSegment<FkTypes.FkPowerSpectra>}>>
 *
 * A record of Fk Channel Segments
 */
export type FkChannelSegmentRecord = Record<
  ChannelSegmentDescriptorId,
  ChannelSegmentTypes.ChannelSegment<FkTypes.FkPowerSpectra>
>;

/**
 * Record<JSON.stringify({@link SignalDetectionIdString}), {@link FkTypes.FkFrequencyThumbnail[]}>>
 *
 * A record of FkFrequencyThumbnails associated to a SignalDetection
 */
export type FkFrequencyThumbnailRecord = Record<
  SignalDetectionIdString,
  FkTypes.FkFrequencyThumbnail[]
>;

/**
 * Record<{@link ChannelName}, {@link ChannelTypes.Channel}>
 *
 * A record of channel names to a matching channel.
 */
export type ChannelRecord = Record<ChannelName, ChannelTypes.Channel>;

/**
 * Record<{@link WeavessRowName}, {@link FilterTypes.Filter}>
 *
 * A record of weavess row names (either a channel name or station name) to a filter.
 */
export type ChannelFilterRecord = Record<WeavessRowName, FilterTypes.Filter>;

/**
 * Record<{@link FilterName}, Record<{@link SampleRate}, {@link FilterDefinition}[]>>
 *
 * A record of filter names, each containing a record of sample rates to a filter definition.
 */
export type FilterDefinitionsRecord = Record<FilterName, Record<SampleRate, FilterDefinition>>;

/**
 * Record<{@link SignalDetectionHypothesisId}, {@link FilterDefinitionByFilterDefinitionUsage}>
 */
export type FilterDefinitionsForSignalDetectionsRecord = Record<
  SignalDetectionHypothesisId,
  FilterDefinitionByFilterDefinitionUsage
>;

/**
 * Record<JSON.stringify({@link ChannelSegmentTypes.ChannelSegmentDescriptor}), {@link FilterDefinitionByFilterDefinitionUsage}>
 */
export type FilterDefinitionsForChannelSegmentsRecord = Record<
  ChannelSegmentDescriptorId,
  FilterDefinitionByFilterDefinitionUsage
>;

/**
 * Record<{@link FilterNameId}, Record<{@link ChannelName}, Set<{@link ChannelSegmentDescriptorId}>>>
 *
 * A record of filter name ids (a combination of the namedFilter name + filter definition name), each
 * containing a record of channel names to a set of unique channel segment descriptor ids
 * (stringified channelSegmentDescriptors).
 */
export type ProcessedItemsCacheRecord = Record<
  FilterNameId,
  Record<ChannelName, Set<ChannelSegmentDescriptorId>>
>;

/**
 * Record<{@link SignalDetectionIdString}, {@link SignalDetectionTypes.SignalDetection}>
 *
 * A record of signal detection ids to signal detections.
 */
export type SignalDetectionsRecord = Record<
  SignalDetectionIdString,
  SignalDetectionTypes.SignalDetection
>;

/**
 * Record<{@link EventId}, {@link EventTypes.Event}>
 *
 * A record of event names to an event.
 */
export type EventsRecord = Record<EventId, EventTypes.Event>;

/**
 * Record<{@link SignalDetectionIdString}, {@link EventTypes.AssociationConflict}>
 *
 * A record of Association Conflicts.
 */
export type AssociationConflictRecord = Record<SignalDetectionIdString, AssociationConflict>;

/**
 * Record<{@link ChannelName}, Record<{@link QcSegmentId}, {@link QcSegment}>>
 *
 * A record of channel names containing a record of qc segment id's to qc segments.
 */
export type QcSegmentRecord = Record<ChannelName, Record<QcSegmentId, QcSegment>>;

/**
 * Used by {@link FilterAssociation} in conjunction
 * with {@link FilterDefinitionAssociationsObject}.
 * Represents the id/startTime of a ChannelSegment that was filtered using a
 * particular {@link FilterTypes.FilterDefinition}
 */
export interface WaveformIdentifier {
  channelSegmentId: string;
  startTime: number;
}

/**
 * Used in conjunction with {@link FilterDefinitionAssociationsObject}.
 * Represents a collection of ChannelSegments that were filtered with a
 * particular {@link FilterTypes.FilterDefinition}
 */
export interface FilterAssociation {
  definition: FilterTypes.FilterDefinition;
  waveformIdentifiers: WaveformIdentifier[];
}

/**
 * This object associates {@link WeavessTypes.DataSegment}s to
 * designed {@link FilterTypes.FilterDefinition} objects for the purpose of file exporting.
 */
export interface FilterDefinitionAssociationsObject {
  filterAssociations: FilterAssociation[];
  channelSegments: UiChannelSegment[];
}

/**
 * Record<{@link ChannelSegmentDescriptorId}, {@link SignalDetectionHypothesisId}>
 *
 * A record of channel segment descriptor strings to signal detection hypothesis id
 */
export type ChannelSegmentsToSignalDetectionHypothesisRecord = Record<
  ChannelSegmentDescriptorId,
  SignalDetectionHypothesisId
>;
