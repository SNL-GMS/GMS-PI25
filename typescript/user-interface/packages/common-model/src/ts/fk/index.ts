export * from './types';
export * as Util from './util';
