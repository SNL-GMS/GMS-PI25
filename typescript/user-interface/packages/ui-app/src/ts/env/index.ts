const ENV: Record<string, string> = Object.freeze((window as any).ENV);

/**
 * Used to determine if an environment variable is not set
 *
 * @param envVar to be tested
 * @returns true when no value is set
 */
export function isNotSet(envVar: string) {
  return (
    envVar === 'undefined' ||
    envVar === undefined ||
    envVar === null ||
    envVar === 'null' ||
    envVar.length === 0
  );
}

const GMS_KEYCLOAK = Object.freeze({
  GMS_KEYCLOAK_REALM: ENV?.GMS_KEYCLOAK_REALM,
  GMS_KEYCLOAK_URL: ENV?.GMS_KEYCLOAK_URL,
  GMS_KEYCLOAK_CLIENT_ID: ENV?.GMS_KEYCLOAK_CLIENT_ID,
  GMS_DISABLE_KEYCLOAK_AUTH: ENV?.GMS_DISABLE_KEYCLOAK_AUTH === 'true'
});

export const {
  GMS_KEYCLOAK_REALM,
  GMS_KEYCLOAK_URL,
  GMS_KEYCLOAK_CLIENT_ID,
  GMS_DISABLE_KEYCLOAK_AUTH
} = GMS_KEYCLOAK;
