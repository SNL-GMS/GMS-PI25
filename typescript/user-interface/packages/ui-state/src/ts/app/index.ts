export * from './api';
export * from './history';
export * from './hooks';
export * from './query';
export * from './redux-provider';
export * from './sanitization';
export * from './state';
export * from './store';
export * from './subscription';
export * from './util';
