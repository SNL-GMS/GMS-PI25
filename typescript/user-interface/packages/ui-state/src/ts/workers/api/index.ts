export * from './compute-fk-spectra';
export * from './export-channel-segments';
export * from './fetch-channel-segments-by-channel';
export * from './fetch-default-filter-definition-by-usage-for-channel-segments';
export * from './fetch-default-filter-definitions-for-signal-detection-hypotheses';
export * from './fetch-events-detections-segments-by-time';
export * from './fetch-filter-definitions-for-signal-detections';
export * from './fetch-signal-detections-segments-by-stations-time';
export * from './get-boundaries';
export * from './get-position-buffer';
