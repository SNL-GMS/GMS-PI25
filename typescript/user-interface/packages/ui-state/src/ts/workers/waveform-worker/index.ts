import { Timer } from '@gms/common-util';
import { UILogger } from '@gms/ui-util';
import { RpcProvider } from 'worker-rpc';

import { cancelWorkerRequests } from './operations/cancel-worker-requests';
import { clearWaveforms } from './operations/clear-waveforms';
import { computeFkSpectra } from './operations/compute-fk-spectra';
import { exportChannelSegmentsWithFilterAssociations } from './operations/export-channel-segments';
import { fetchChannelSegmentsByChannel } from './operations/fetch-channel-segments-by-channel';
import { fetchChannelsByNamesTimeRange } from './operations/fetch-channels-by-names-timerange';
import { fetchDefaultFilterDefinitionByUsageForChannelSegments } from './operations/fetch-default-filter-definition-by-usage-for-channel-segments';
import { fetchDefaultFilterDefinitionsForSignalDetectionHypotheses } from './operations/fetch-default-filter-definitions-for-signal-detection-hypotheses';
import { fetchEventsAndDetectionsWithSegments } from './operations/fetch-events-detections-segments-by-time';
import { fetchFilterDefinitionsForSignalDetections } from './operations/fetch-filter-definitions-for-signal-detections';
import { fetchSignalDetectionsWithSegments } from './operations/fetch-signal-detections-segments-by-stations-time';
import { getBoundaries } from './operations/get-boundaries';
import { getWaveform } from './operations/get-waveform';
import { WorkerOperations } from './operations/operations';
import {
  designFilter,
  filterChannelSegment,
  filterChannelSegments
} from './operations/ui-filter-processor';

const logger = UILogger.create('GMS_LOG_RPC_PROVIDER', process.env.GMS_LOG_RPC_PROVIDER);

/**
 * Set this to true to use transfer objects (this will cause the worker to lose access to that data).
 * Not yet supported.
 */
const SHOULD_USE_TRANSFER = true;

/**
 * Handle the message passed to the worker. Dispatch the requested operation.
 */
// eslint-disable-next-line no-restricted-globals
((self as unknown) as SharedWorkerGlobalScope).onconnect = connectEvent => {
  const port = connectEvent.ports[0];

  /**
   * Create the rpcProvider and tell it how to send messages.
   * TODO: remove RpcProvider as it is no longer needed for worker compatibility with webpack
   * TODO: will need to re-write registration functions
   */
  const rpcProvider = new RpcProvider((message, transfer: any[]) => {
    Timer.end('[Waveform worker]', 1000);

    if (SHOULD_USE_TRANSFER && ArrayBuffer.isView(message.payload)) {
      // We can only use transferObjects for ArrayBuffers and their Views (such as Float32Array)
      port.postMessage(message, [message.payload.buffer]);
    } else {
      // Copy message data via structured cloning
      port.postMessage(message, transfer);
    }
  });

  /**
   * The error event is dispatched if there is either a local or remote communication error (timeout, invalid id, etc.)
   * This will also catch if you forgot to register the worker operation :)
   */
  rpcProvider.error.addHandler(err => {
    logger.error(err.message);
  });

  /** RPC Handler Registration */
  rpcProvider.registerRpcHandler(WorkerOperations.COMPUTE_FK_SPECTRA, computeFkSpectra);
  rpcProvider.registerRpcHandler(
    WorkerOperations.FETCH_SIGNAL_DETECTIONS_WITH_SEGMENTS_BY_STATIONS_TIME,
    fetchSignalDetectionsWithSegments
  );
  rpcProvider.registerRpcHandler(
    WorkerOperations.FETCH_CHANNEL_SEGMENTS_BY_CHANNEL,
    fetchChannelSegmentsByChannel
  );
  rpcProvider.registerRpcHandler(
    WorkerOperations.FETCH_EVENTS_WITH_DETECTIONS_AND_SEGMENTS_BY_TIME,
    fetchEventsAndDetectionsWithSegments
  );
  rpcProvider.registerRpcHandler(WorkerOperations.GET_WAVEFORM, getWaveform);
  rpcProvider.registerRpcHandler(WorkerOperations.GET_BOUNDARIES, getBoundaries);
  rpcProvider.registerRpcHandler(WorkerOperations.CLEAR_WAVEFORMS, clearWaveforms);
  rpcProvider.registerRpcHandler(WorkerOperations.CANCEL_WORKER_REQUESTS, cancelWorkerRequests);
  rpcProvider.registerRpcHandler(WorkerOperations.FILTER_CHANNEL_SEGMENTS, filterChannelSegments);
  rpcProvider.registerRpcHandler(WorkerOperations.FILTER_CHANNEL_SEGMENT, filterChannelSegment);
  rpcProvider.registerRpcHandler(WorkerOperations.DESIGN_FILTER, designFilter);
  rpcProvider.registerRpcHandler(
    WorkerOperations.EXPORT_CHANNEL_SEGMENTS,
    exportChannelSegmentsWithFilterAssociations
  );
  rpcProvider.registerRpcHandler(
    WorkerOperations.FETCH_CHANNELS_BY_NAMES_TIME_RANGE,
    fetchChannelsByNamesTimeRange
  );
  rpcProvider.registerRpcHandler(
    WorkerOperations.FETCH_FILTER_DEFINITIONS_FOR_SIGNAL_DETECTIONS,
    fetchFilterDefinitionsForSignalDetections
  );
  rpcProvider.registerRpcHandler(
    WorkerOperations.FETCH_DEFAULT_FILTER_DEFINITIONS_FOR_SIGNAL_DETECTION_HYPOTHESES,
    fetchDefaultFilterDefinitionsForSignalDetectionHypotheses
  );
  rpcProvider.registerRpcHandler(
    WorkerOperations.FETCH_DEFAULT_FILTER_DEFINITION_BY_USAGE_FOR_CHANNEL_SEGMENTS,
    fetchDefaultFilterDefinitionByUsageForChannelSegments
  );

  /**
   * Handle the message passed to the worker. Dispatch the requested operation.
   */
  port.onmessage = messageEvent => {
    Timer.start('[Waveform worker]');
    rpcProvider.dispatch(messageEvent.data);
  };
};
