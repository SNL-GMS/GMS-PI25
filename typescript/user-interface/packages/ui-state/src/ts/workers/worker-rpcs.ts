import { RpcProvider } from 'worker-rpc';

// To debug the worker use chrome://inspect/#workers
export const waveformWorker = new SharedWorker(
  new URL('./waveform-worker/index.ts', import.meta.url) /* webpackChunkName: 'waveform-worker' */,
  'waveform-worker'
);

// TODO remove RpcProvider as it is no longer needed for worker compatibility with webpack
export const waveformWorkerRpc = new RpcProvider((message, transfer) => {
  waveformWorker.port.postMessage(message, transfer);
});
waveformWorker.port.onmessage = e => {
  waveformWorkerRpc.dispatch(e.data);
};
