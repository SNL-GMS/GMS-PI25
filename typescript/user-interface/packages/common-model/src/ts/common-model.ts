import * as BeamformingTemplateTypes from './beamforming-templates/types';
import * as ChannelSegmentTypes from './channel-segment/types';
import * as ColorTypes from './color/types';
import * as CommonTypes from './common/types';
import * as DataAcquisitionTypes from './data-acquisition/types';
import * as Displays from './displays/types';
import * as Endpoints from './endpoints/types';
import * as EventTypes from './event';
// required for legacy components
import * as LegacyEventTypes from './event/legacy';
import * as FacetedTypes from './faceted';
import * as FilterUtil from './filter/filter-util';
import * as FilterTypes from './filter/types';
import * as FkTypes from './fk';
import * as ProcessingMaskDefinitionTypes from './processing-mask-definitions/types';
// required for legacy components
import * as QcMaskTypes from './qc-mask';
import * as QcSegmentTypes from './qc-segment';
import * as SignalDetectionTypes from './signal-detection';
import * as SohTypes from './soh/types';
import * as ChannelTypes from './station-definitions/channel-definitions/channel-definitions';
import * as ResponseTypes from './station-definitions/channel-definitions/response-definitions';
import * as StationTypes from './station-definitions/station-definitions/station-definitions';
import * as ProcessingStationTypes from './station-processing/types';
import * as ReferenceStationTypes from './station-reference/types';
import * as SystemMessageTypes from './system-message/types';
import * as TimeTypes from './time';
import * as TypeUtil from './type-util/type-util';
import * as ConfigurationTypes from './ui-configuration/types';
import * as UserProfileTypes from './user-profile/types';
import * as WaveformTypes from './waveform/types';
import * as WorkflowTypes from './workflow/types';

export {
  BeamformingTemplateTypes,
  ChannelSegmentTypes,
  ChannelTypes,
  ColorTypes,
  CommonTypes,
  ConfigurationTypes,
  DataAcquisitionTypes,
  Displays,
  Endpoints,
  EventTypes,
  FacetedTypes,
  FilterTypes,
  FilterUtil,
  FkTypes,
  LegacyEventTypes,
  ProcessingMaskDefinitionTypes,
  ProcessingStationTypes,
  QcMaskTypes,
  QcSegmentTypes,
  ReferenceStationTypes,
  ResponseTypes,
  SignalDetectionTypes,
  SohTypes,
  StationTypes,
  SystemMessageTypes,
  TimeTypes,
  TypeUtil,
  UserProfileTypes,
  WaveformTypes,
  WorkflowTypes
};
