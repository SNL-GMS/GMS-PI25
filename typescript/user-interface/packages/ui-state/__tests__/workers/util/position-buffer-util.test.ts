/* eslint-disable @typescript-eslint/no-magic-numbers */
import { linearFilterDefinition } from '@gms/common-model/__tests__/__data__';
import type { WeavessTypes } from '@gms/weavess-core';

import { WaveformStore } from '../../../src/ts/workers';
import {
  changeUniqueIdFilter,
  convertAndStoreTimeseries
} from '../../../src/ts/workers/waveform-worker/util/position-buffer-util';
import { channelSegmentWithSamples, claimCheckChannelSegment } from '../../__data__';

const domain: WeavessTypes.TimeRange = {
  startTimeSecs: channelSegmentWithSamples.timeseries[0].startTime,
  endTimeSecs: channelSegmentWithSamples.timeseries[0].endTime
};
describe('Position Buffer Utils', () => {
  it('has a WaveformStore it can access', () => {
    expect(WaveformStore).toBeDefined();
  });

  it('convertAndStoreTimeseries', async () => {
    const result = await convertAndStoreTimeseries(channelSegmentWithSamples, domain);
    expect(result).toMatchSnapshot();
    expect(await WaveformStore.retrieve(result[0]._uiClaimCheckId)).toBeDefined();
  });

  it('changeUniqueIdFilter', () => {
    expect(() => changeUniqueIdFilter(undefined)).toThrow(
      'There was an error parsing the uniqueId'
    );
    expect(
      claimCheckChannelSegment.timeseries[0]._uiClaimCheckId.includes(linearFilterDefinition.name)
    ).toBeFalsy();
    const changedId = changeUniqueIdFilter(
      claimCheckChannelSegment.timeseries[0]._uiClaimCheckId,
      linearFilterDefinition.name
    );
    expect(changedId.includes(linearFilterDefinition.name)).toBeTruthy();
  });
});
