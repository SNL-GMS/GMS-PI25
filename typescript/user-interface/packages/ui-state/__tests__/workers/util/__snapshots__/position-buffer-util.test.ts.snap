// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Position Buffer Utils convertAndStoreTimeseries 1`] = `
[
  {
    "_uiClaimCheckId": "{"domain":{"startTimeSecs":1638297900,"endTimeSecs":1638298200},"id":{"channel":{"effectiveAt":1638297900,"name":"PDAR.PD01.SHZ"},"startTime":1638297900,"endTime":1638298200,"creationTime":1638297900},"type":"WAVEFORM","filter":"Unfiltered","waveform":{"type":"WAVEFORM","startTime":1638297900,"endTime":1638298200,"sampleCount":12000,"sampleRateHz":40}}",
    "endTime": 1638298200,
    "sampleCount": 12000,
    "sampleRateHz": 40,
    "samples": undefined,
    "startTime": 1638297900,
    "type": "WAVEFORM",
  },
]
`;
